### PR TITLE
Transaction Batches Stateful Validation

### PR DESCRIPTION
**Pull request by Akvinikym**
### Description of the Change

Adds stateful validation checks for transaction batches:

1. If the transaction does not belong to any batch, proceed as usual
2. If the transaction belongs to ordered batch, in case of failure just move to the next transaction again
3. If the transaction belongs to atomic batch, failure of it leads to failure of all transactions in that batch

Also, tests for stateful validator and transaction batch validation were added.

### Benefits

Logic of transaction batches is improved.

### Possible Drawbacks 

Additional overhead for validation and WSV save points management.

### Alternative designs

Savepoint wrapper's design can be a theme of discussion. Shortly, I have decided to implement it as a nested structure of TemporaryWsv, because:
1. It cannot be used without temp WSV - there's just no sense in it
2. It must "see" some of private fields of temp WSV in order to function

There are two alternatives I see: declare it in some kind of temp WSV common file, and then either parametrize it with a template, or make necessary field of temp WSV friends to the wrapper. Both solutions, in my opinion, looks worse, than the one implemented

### CODE REVIEW

File: irohad/ametsuchi/impl/temporary_wsv_impl.cpp [hunk 1]
```diff
@@ -78,16 +78,28 @@ namespace iroha {
...
           }
         }
         // success
-        transaction_->exec("RELEASE SAVEPOINT savepoint_;");
+        releaseSavepoint("savepoint_");
         return {};
       };
     }
 
+    void TemporaryWsvImpl::createSavepoint(const std::string &name) {
```
Comment (l4l) [inline irohad/ametsuchi/impl/temporary_wsv_impl.cpp]: It's barely safe. Return please new object with RAII behavior, such that dtor calls rollback, unless specified explicit release

File: irohad/ametsuchi/impl/temporary_wsv_impl.hpp [hunk 2]
```diff
@@ -30,6 +30,20 @@ namespace iroha {
   namespace ametsuchi {
     class TemporaryWsvImpl : public TemporaryWsv {
      public:
+      struct SavepointWrapper : public TemporaryWsv::SavepointWrapper {
```
Comment (kamilsa) [inline irohad/ametsuchi/impl/temporary_wsv_impl.hpp]: Rename it to `SavepointWrapperImpl`

File: irohad/ametsuchi/impl/temporary_wsv_impl.hpp [hunk 3]
```diff
@@ -39,17 +53,14 @@ namespace iroha {
               const shared_model::interface::Transaction &, WsvQuery &)>
               function) override;
 
-      void createSavepoint(const std::string &name) override ;
-
-      void releaseSavepoint(const std::string &name) override ;
-
-      void rollbackToSavepoint(const std::string &name) override;
+      std::shared_ptr<TemporaryWsv::SavepointWrapper> createSavepoint(
```
Comment (kamilsa) [inline irohad/ametsuchi/impl/temporary_wsv_impl.hpp]: Better to return `unique_ptr`

File: irohad/ametsuchi/temporary_wsv.hpp [hunk 4]
```diff
@@ -39,6 +39,19 @@ namespace iroha {
...
      public:
+      /**
+       * Wrapper for savepoints in wsv state; rollbacks to savepoint, if
+       * destroyed without explicit release, releases it otherwise
+       */
+      struct SavepointWrapper {
+        /**
+         * Release the savepoint
+         */
+        virtual void release() = 0;
```
Comment (kamilsa) [inline irohad/ametsuchi/temporary_wsv.hpp]: Can we put release logic into destructor?

File: irohad/ametsuchi/temporary_wsv.hpp [hunk 5]
```diff
@@ -56,24 +69,14 @@ namespace iroha {
...
-       * @param name of savepoint to be rollbacked to
-       */
-      virtual void rollbackToSavepoint(const std::string &name) = 0;
+      virtual std::shared_ptr<SavepointWrapper> createSavepoint(const std::string &name) = 0;
 
       virtual ~TemporaryWsv() = default;
-    };
-  }  // namespace ametsuchi
+
+    };  // namespace ametsuchi
```
Comment (kamilsa) [inline irohad/ametsuchi/temporary_wsv.hpp]: Fix namespace duplication

File: irohad/ametsuchi/temporary_wsv.hpp (hunk 4 above)
Comment (l4l) [reply irohad/ametsuchi/temporary_wsv.hpp]: I don't think that's a good idea. It completely implementation specific stuff

Comment (Akvinikym) [reply irohad/ametsuchi/temporary_wsv.hpp]: Yeah, we've already spoken about this with Kamil, decided to leave it as-is

**Pull request merged**